Mathlib/Tactic/Positivity/Basic.lean

```diff
@@ -328,28 +330,31 @@ Since the output type of `Int.natAbs` is `ℕ`, the nonnegative case is handled
 `positivity` tactic.
 -/
 @[positivity Int.natAbs _]
-def evalNatAbs : PositivityExt where eval {_u _α} _zα _pα (e : Q(ℕ)) := do
-  let ~q(Int.natAbs $a) := e | throwError "not Int.natAbs"
-  let zα' : Q(Zero Int) := q(inferInstance)
-  let pα' : Q(PartialOrder Int) := q(inferInstance)
-  let ra ← core zα' pα' a
-  match ra with
-  | .positive pa =>
-    pure (.positive (q(int_natAbs_pos $pa) : Expr))
-  | .nonzero pa =>
-    pure (.positive (q(Int.natAbs_pos.mpr $pa) : Expr))
-  | .nonnegative _pa =>
-    pure .none
-  | .none =>
-    pure .none
+def evalNatAbs : PositivityExt where eval {u α} _zα _pα e := do
+  match u, α, e with
+  | 0, ~q(ℕ), ~q(Int.natAbs $a) =>
+    let zα' : Q(Zero Int) := q(inferInstance)
+    let pα' : Q(PartialOrder Int) := q(inferInstance)
+    let ra ← core zα' pα' a
+    match ra with
+    | .positive pa =>
+      assertInstancesCommute
```

Comment:
Sometimes you write `assertInstancesCommute` before the `match`, sometimes after. Is there a reason for this (that you can document somewhere easily found)?

Reply:
All that matters is the position of the `assertInstancesCommute`/`assumeInstancesCommute` wrt the `inferInstance`s. In fact, all that matters is that there's no `inferInstance` not followed by `assertInstancesCommute`/`assumeInstancesCommute` before a `q()` quotation. This is something that's automatically checked by the Qq quotations as you write the code.